File: python/901.py
from math import inf


class StockSpanner:

    def __init__(self):
        self.stack = [(-1, inf)]
        self.cur_day = -1

    def next(self, price: int) -> int:
        while self.stack[-1][1] <= price:
            self.stack.pop()
        self.cur_day += 1
        self.stack.append((self.cur_day, price))
        return self.cur_day - self.stack[-2][0]



# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: go/901.go
package main

import "math"

type pair901 struct {
	day   int
	price int
}

type StockSpanner struct {
	stack  []pair901
	curDay int
}

func Constructor901() StockSpanner {
	return StockSpanner{[]pair901{{-1, math.MaxInt}}, -1}
}

func (this *StockSpanner) Next(price int) int {
	for price >= this.stack[len(this.stack)-1].price {
		this.stack = this.stack[:len(this.stack)-1]
	}
	this.curDay++
	this.stack = append(this.stack, pair901{this.curDay, price})
	return this.curDay - this.stack[len(this.stack)-2].day
}
